### PR TITLE
using existing connection for scripting as operation 

### PR DESCRIPTION
**Pull request by llali**
- Getting connection from binding queue to be able to use existing connection fr script as operations
- Getting existing connection from language service when getting object definition
-Removed the duplicate options from ScriptingOption public class, Created a wrapper for ScriptingOption to be able to add the duplicate option names for Script as operation.

### CODE REVIEW

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsOptions.cs [hunk 1]
```diff
@@ -0,0 +1,712 @@
...
+//
+
+using Microsoft.SqlTools.ServiceLayer.Scripting.Contracts;
+using Microsoft.SqlTools.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer.Scripting
+{
+    /// <summary>
+    /// The options for Script as operation
+    /// </summary>
```
Comment (pensivebrian) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsOptions.cs]: Can you add a comment that this class is mainly an adapter to map property names?

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsOptions.cs [hunk 2]
```diff
@@ -0,0 +1,712 @@
...
+
+        /// <summary>
+        /// Script only features compatible with the specified version of SQL Server.  Possible values:
+        ///   Script90Compat
+        ///   Script100Compat
+        ///   Script105Compat
+        ///   Script110Compat
+        ///   Script120Compat
+        ///   Script130Compat
+        ///   Script140Compat
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsOptions.cs]: Script150Compat should be present too... right? Or is that not available yet...?

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsOptions.cs [hunk 3]
```diff
@@ -0,0 +1,712 @@
...
+        /// <summary>
+        /// Script only features compatible with the specified SQL Server database engine edition.
+        /// Possible Values:
+        ///   SqlServerPersonalEdition
+        ///   SqlServerStandardEdition 
+        ///   SqlServerEnterpriseEdition 
+        ///   SqlServerExpressEdition
+        ///   SqlAzureDatabaseEdition
+        ///   SqlDatawarehouseEdition
+        ///   SqlServerStretchEdition
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsOptions.cs]: SqlManagedInstanceEdition also?

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsScriptingOperation.cs [hunk 4]
```diff
@@ -41,28 +48,26 @@ public override void Execute()
...
-                    options.IncludeHeaders = false;
-                    scripter.Options = options;
-                    scripter.Options.ScriptData = false;
-                    scripter.ScriptingError += ScripterScriptingError;
-                    UrnCollection urns = CreateUrns(serverConnection);
-                    var result = scripter.Script(urns);
-                    resultScript = GetScript(options, result);
-                }
+
+                Server server = new Server(ServerConnection);
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsScriptingOperation.cs]: What happens if the connection is closed? Will this handle auto-reopening it or do we need to cover that ourselves? We have cases where the connection is closed and in query execution we now auto reopen before doing the operation

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingService.cs [hunk 5]
```diff
@@ -155,10 +151,34 @@ public void InitializeService(ServiceHost serviceHost)
...
-        public async Task HandleScriptExecuteRequest(ScriptingParams parameters, RequestContext<ScriptingResult> requestContext)
+        private void RunScriptAsTask(ConnectionInfo connInfo, ScriptingParams parameters, RequestContext<ScriptingResult> requestContext)
+        {
+            ScriptAsScriptingOperation operation = new ScriptAsScriptingOperation(parameters);
+            ConnectionServiceInstance.ConnectionQueue.QueueBindingOperation(
+                key: ConnectionServiceInstance.ConnectionQueue.AddConnectionContext(connInfo, "Scripting"),
+                bindingTimeout: ScriptingOperationTimeout,
+                bindOperation: (bindingContext, cancelToken) =>
+                {
+                    string script = string.Empty;
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingService.cs]: this seems unused? Can you remove?

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsOptions.cs (hunk 2 above)
Comment (llali) [reply src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsOptions.cs]: not sure if it's available, I don't see it as a valid option in SM. I'll verify ylater

---
In reply to: [154440666](https://github.com/Microsoft/sqltoolsservice/pull/556#discussion_r154440666) [](ancestors = 154440666)

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingService.cs (hunk 5 above)
Comment (llali) [reply src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingService.cs]: I'm changing this code in another PR 

---
In reply to: [154441355](https://github.com/Microsoft/sqltoolsservice/pull/556#discussion_r154441355) [](ancestors = 154441355)

File: src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingService.cs [hunk 6]
```diff
@@ -155,10 +151,34 @@ public void InitializeService(ServiceHost serviceHost)
...
+                    operation.ServerConnection = bindingContext.ServerConnection;
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingService.cs]: this is a bit of a problem in that we are taking values from bindingContext and saving them to a variable that survives outside the locked action section.  this isn't really a good design practice as this context should only be used within the locked callback functions.

**Pull request merged**